### ISSUE TITLE
DPG replaces existing ServiceVersion values in ClientOptions

### ISSUE THREAD

**Issue by heaths** [Azure/autorest.csharp#2391]
The DPG-generated `*ClientOptions.ServiceVersion` enum only ever has 1 version defined as value 1. This is a problem for CLU since we just shipped a 1.0 GA with `V2022_05_01 = 1` and now need to ship a CLU 1.1 beta with `V2022_05_15_Preview = 2` (note also that the casing is wrong by default; see #1524).

After a lot of trial and error as well as some offline advice, I found I could work around this for now using `[CodeGenSuppressType("ConversationsClientOptions")]` and just author the whole thing myself, but since SDKs will often have to support new versions this is something we'll want to handle better, I hope.

**Comment by archerzz** [comment]
CADL supports multiple API version. Considering the current situation:
- We'll move to CADL
- There are not many swagger-based DPG SDKs

I suggest:
- We support multiple API in CADL
- Use customization codes for current DPG SDKs

@heaths  @annelo-msft Any thought? Thanks.

**Comment by heaths** [comment]
I'm hopeful, but skeptical that this won't be a problem before we can use CADL. Cognitive Language Services is already finishing up another preview for which we'll need a beta (plus what we already GA'd) and another preview right around the corner.

I think we at least need the generator to stop generating the ServiceVersion if we define it. Perhaps assume a `LatestVersion` forcing us to use it, which I think is fine.

**Comment by annelo-msft** [comment]
We are tracking work on multi-version support with this issue: https://github.com/Azure/autorest.csharp/issues/2362, and plan to address it after our August MVP.

**Comment by heaths** [comment]
@annelo-msft my concern isn't just about when the feature is supported, but when the services are migrated. We already have the code to avoid generating types and members when they are already defined. Why can't we just use that for the nested `ServiceVersion` enum as well? Doesn't have to be fancy: if it's defined, just don't define it again. I think it's perfectly reasonable to expect devs to completely take over the enum definition themselves in this case (no such thing as partial enums anyway).

**Comment by archerzz** [comment]
> I'm hopeful, but skeptical that this won't be a problem before we can use CADL. Cognitive Language Services is already finishing up another preview for which we'll need a beta (plus what we already GA'd) and another preview right around the corner.
> 
> I think we at least need the generator to stop generating the ServiceVersion if we define it. Perhaps assume a `LatestVersion` forcing us to use it, which I think is fine.

You can write a customized `*ClientOptions` to define multiple `ServiceVersion`. Some SDKs are already doing this in HLC, like [Azure.Storage.Blobs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/src/BlobClientOptions.cs).

**Comment by heaths** [comment]
How is that supposed to work with generated DPG clients, though? If you hand-code a client you can do whatever with client options, but if the idea - the goal - is to generate DPG clients as the public API customers should use, I've not been able to define a partial ClientOptions that the DPG client uses without having the ServiceVersion redefined as public.

I'm aware I can define a `SeviceVersion` enum with multiple values. We do that in just about every SDK and could with HLC/older LLC, but can't with DPG. That's actually why I opened this bug. Because code gen *always* generates `enum ServiceVersion` and `const ServiceVersion LatestVersion`, I can't customize a client options class to overwrite these. That's what I want.

**Comment by jsquire** [comment]
I agree with Heath's perspective; this is a very real need and will be a common scenario as we onboard new DPG libraries and evolve those that we have.   I understand that we may not have a complete end-to-end story around this until after the August MVP, but we do need a viable work-around in the short term.

@archerzz, @AlexanderSher:  I'd appreciate any thoughts that you have.

**Comment by heaths** [comment]
Just to be clear, I'm not asking for anything fancy. I appreciate a solution may be short lived (by some definition of "short", but still probably months out). We already have code to avoid generating pre-defined types and/or members, so we could utilize that. If one or both of `ServiceVersion` or `LatestVersion` is defined, generate neither: force devs to define both and just assume they are there in the generated code. No need to try to add values as the latest integer to `ServiceVersion` or figure out what that latest version is to define `LatetVersion`. In C/C++ parlance, assume they are `extern` definitions and will be defined. If compile issues occur after, that's on the dev to define both if they only defined one. I hope this would be an easy solution as it puts the responsibility on the SDK dev.
